Lib9c/Model/WorldInformation.cs

```diff
@@ -351,55 +351,74 @@ public bool TryGetLastClearedStageId(out int stageId)
         /// <param name="clearedAt"></param>
-        /// <param name="unlockSheet"></param>
+        /// <param name="worldSheet"></param>
+        /// <param name="worldUnlockSheet"></param>
+        /// <exception cref="FailedToUnlockWorldException"></exception>
         public void ClearStage(
             int worldId,
             int stageId,
             long clearedAt,
-            WorldUnlockSheet unlockSheet)
+            WorldSheet worldSheet,
+            WorldUnlockSheet worldUnlockSheet)
         {
+            // NOTE: Always consider world unlock.
+            // Because even a stage that has already been cleared can be a trigger for world unlock due to the table patch.
+            if (worldUnlockSheet.TryGetUnlockedInformation(worldId, stageId, out var worldIdsToUnlock))
+            {
+                foreach (var worldIdToUnlock in worldIdsToUnlock)
+                {
+                    UnlockWorld(worldIdToUnlock, clearedAt, worldSheet);
+                }
+            }
+
             var world = _worlds[worldId];
             if (stageId <= world.StageClearedId)
             {
                 return;
             }
 
             _worlds[worldId] = new World(world, clearedAt, stageId);
-
-            if (unlockSheet.TryGetUnlockedInformation(worldId, stageId, out var worldIdsToUnlock))
-            {
-                foreach (var worldIdToUnlock in worldIdsToUnlock)
-                {
-                    UnlockWorld(worldIdToUnlock, clearedAt);
-                }
-            }
         }
 
         /// <summary>
-        /// 특정 월드를 잠금 해제한다.
+        /// Unlock a specific world.
         /// </summary>
         /// <param name="worldId"></param>
         /// <param name="unlockedAt"></param>
-        /// <exception cref="KeyNotFoundException"></exception>
-        private void UnlockWorld(int worldId, long unlockedAt)
+        /// <param name="worldSheet"></param>
+        /// <exception cref="FailedToUnlockWorldException"></exception>
+        private void UnlockWorld(int worldId, long unlockedAt, WorldSheet worldSheet)
         {
-            if (!_worlds.ContainsKey(worldId))
+            World world;
+            if (_worlds.ContainsKey(worldId))
             {
-                throw new KeyNotFoundException($"{nameof(worldId)}: {worldId}");
+                world = _worlds[worldId];
+            }
+            else if (!worldSheet.TryGetValue(worldId, out var worldRow) ||
+                     !TryAddWorld(worldRow, out world))
+            {
+                throw new FailedToUnlockWorldException($"{nameof(worldId)}: {worldId}");
             }
 
-            if (_worlds[worldId].IsUnlocked)
+            if (world.IsUnlocked)
             {
                 return;
             }
 
-            var world = _worlds[worldId];
             _worlds[worldId] = new World(world, unlockedAt);
         }
     }
+
+    [Serializable]
+    public class FailedToUnlockWorldException : Exception
```

Comment:
I guess we need to add `protected FailedToUnlockWorldException(SerializationInfo info, StreamingContext context) : base(info, context)` for .NET Serializable. (see also https://stackoverflow.com/questions/94488/what-is-the-correct-way-to-make-a-custom-net-exception-serializable)